compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/DecomposeLinalgExtPackUnPackToAIR.cpp

```diff
@@ -0,0 +1,864 @@
+    result.push_back(getMixedSize(builder, loc, value, i));
+  return result;
+}
+
+// Normalizes a permutation on a higher rank space to its actual size, e.g.
+//   perm = [1, 4, 2]
+// becomes
+//   norm = [0, 2, 1]
+static SmallVector<int64_t> getPackUnpackNormalizedPerm(
+    int rank, ArrayRef<int64_t> perm) {
+  constexpr int64_t kNonTiledMarker = -1;
+  SmallVector<int64_t> vec(rank, kNonTiledMarker);
+  for (auto [index, value] : llvm::enumerate(perm)) vec[value] = index;
+  SmallVector<int64_t> normalizedPerm = llvm::to_vector(llvm::make_filter_range(
+      vec, [&](int64_t v) { return v != kNonTiledMarker; }));
+  // This inverts the permutation in addition to normalizing so invert back.
+  return invertPermutationVector(normalizedPerm);
+}
+
+// Gets the normalized permutation implied by innerDimsPos and outerDimsPerm
+// assuming rank reduction of unit outer dims.
+static SmallVector<int64_t> getPackUnpackRankReducedPerm(
+    ArrayRef<int64_t> shape, ArrayRef<int64_t> innerDimsPos,
+    ArrayRef<int64_t> outerDimsPerm) {
+  SmallVector<int64_t> rankReducedOuterDimsPerm;
+  SmallVector<int64_t> outerDims;
+  SmallVector<int64_t> innerDims;
+  int64_t dim = 0;
+  int64_t unpackedRank = shape.size();
+  for (auto i : llvm::seq<unsigned>(0, unpackedRank)) {
+    if (llvm::is_contained(innerDimsPos, i)) {
+      innerDims.push_back(dim++);
+      continue;
+    }
+    outerDims.push_back(dim++);
+    if (!outerDimsPerm.empty())
+      rankReducedOuterDimsPerm.push_back(outerDimsPerm[i]);
+  }
+
+  // Get the position of the inner dims after permutation.
+  SmallVector<int64_t> innerPerm =
+      getPackUnpackNormalizedPerm(unpackedRank, innerDimsPos);
+  applyPermutationToVector<int64_t>(innerDims, innerPerm);
+
+  // Ditto for the outer dims.
+  SmallVector<int64_t> perm = outerDims;
+
+  rankReducedOuterDimsPerm =
+      getPackUnpackNormalizedPerm(unpackedRank, rankReducedOuterDimsPerm);
+  if (!rankReducedOuterDimsPerm.empty())
+    applyPermutationToVector<int64_t>(perm, rankReducedOuterDimsPerm);
+
+  // The tile always ends up as the inner most dims after packing.
+  perm.append(innerDims);
+
+  return perm;
+}
+
+class GeneralizePackOpPattern
+    : public OpRewritePattern<IREE::LinalgExt::PackOp> {
+ public:
+  using OpRewritePattern<IREE::LinalgExt::PackOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(IREE::LinalgExt::PackOp packOp,
+                                PatternRewriter &rewriter) const override {
+    if (llvm::any_of(packOp.getMixedTiles(),
+                     [](OpFoldResult tile) { return tile.is<Value>(); })) {
+      return rewriter.notifyMatchFailure(
+          packOp, "require inner tile sizes being static");
+    }
+
+    auto innerDimsPos = packOp.getInnerDimsPos();
+    int64_t srcRank = packOp.getInputRank();
+    auto destShape = packOp.getOutputType().getShape();
+    if (llvm::any_of(innerDimsPos, [destShape](int64_t index) {
+          return destShape[index] != 1;
```

Comment:
I dont think this is enough. The outer dimensions could be permuted as well. AFAICS you are looking for cases where the outer pack dimension is 1. So you will need to account for `outer_dims_perm` as well?

Reply:
The `outer_dims_perm` will be handled in `lowerPack` pattern. that happens after this `GeneralizePackOpPattern`.

Reply:
Agreed. Done.